include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -306,6 +306,82 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+
+
+template<int dim, int spacedim>
+std::vector<std::shared_ptr<const Triangulation<dim, spacedim>>>
+create_geometric_coarsening_sequence(const Triangulation<dim, spacedim> & fine_triangulation_in)
+{
+  std::vector<std::shared_ptr<const Triangulation<dim, spacedim>>> coarse_grid_triangulations(
+    fine_triangulation_in.n_global_levels());
+
+  coarse_grid_triangulations.back().reset(&fine_triangulation_in, [](auto *) {
+    // empty deleter, since fine_triangulation_in is an external field
+    // and its destructor is called somewhere else
+  });
+
+  // for a single level nothing has to be done
+  if(fine_triangulation_in.n_global_levels() == 1)
+    return coarse_grid_triangulations;
+
+#ifndef DEAL_II_WITH_P4EST
+  Assert(false, ExcNotImplemented());
+#else
+  const auto fine_triangulation =
+    dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim> *>(
+      &fine_triangulation_in);
+
+  Assert(fine_triangulation, ExcNotImplemented());
+
+  const unsigned int max_level = fine_triangulation->n_global_levels() - 1;
+
+  const auto coarse_mesh_description = GridTools::get_coarse_mesh_description(*fine_triangulation);
```

Comment:
this line can be shifted between lines 344, 345?

Reply:
did you remove this, and why?

Reply:
Because it is done within `temp_tria.copy_triangulation(*fine_triangulation);` on current version of deal.II.